loader/0032-loader-plugins.md

```diff
@@ -0,0 +1,328 @@
+mod launcher as a loader plugin.
+
+## Application
+Loader plugins are initially added from the classpath, allowing default loader
+plugins for an instance to be added via the profile json. All plugins are then
+applied in cycles. During each cycle, the following things happen:
+- Each plugin is given a chance to add any plugin candidates
+    - Each new plugin added this way is given a chance to add any plugin candidates.
+      Rinse and repeat.
+- Each plugin is given a chance to add any required or tentative mod candidates,
+  rules, or options
+- Quilt Loader will evaluate the sat4j tree
+    - Each plugin will be given a chance to claim unsatisfied rules for resolution
+    - Each plugins resolvers will be run sequentially
+
+If any changes to the mod set have taken place during a given cycle, another cycle
+will be run. Otherwise, the game will either launch or crash, depending on whether
+or not all rules have been satisfied.
+
+
+## API Surface
+
+Below is a tentative plan of what the surface level API classes might look like.
+Everything here is subject (and likely) to change, but helps provide a visual
+idea of how developers might use this feature down the road.
+
+<details>
+    <summary>QuiltPluginContext</summary>
+
+```java
+/**
+* Passed to loader plugins to define what actions they are able to take.
+*/
+public sealed interface QuiltPluginContext permits QuiltPluginContextImpl {
+/**
+* The plugin that this context is for. This method is useless, it just indicates that every other method here has an 
+* implicit paramater of "The Loader Plugin" for the UI / logging to use in some way
+*/
+QuiltLoaderPlugin plugin();
+
+void addCandidate(ModCandidate candidate);
+
+void addCandidate(PluginCandidate candidate);
+
+/**
+* Adds a tentative mod candidate which indicates that downloading / fetching a new mod will fix a rule somewhere.
+* This tentative mod won't be kept around to the next cycle - instead the resolver is called to actually download
+* the mod if {@link QuiltLoaderPlugin#canResolve} returns true after each plugin has been checked.
+*/
+void addTentativeCandidate(String group, String modId, Version version, CandidateResolver resolver);
+
+/**
+* Adds a rule to the current solver.
+*/
+void addRule(Rule rule);
+
+/**
+* Adds a LoadOption to the current solver. All existing rules will have Rule#onLoadOptionAdded called, and all plugins 
+* will have ILoaderPlugin#onLoadOptionAdded called.
+*
+* If this is an ITentativeLoadOption then it will be removed at the end of the cycle, and handled by whatever plugin 
+* added it.
+*/
+void addOption(LoadOption option);
+
+/**
+* Gets the metadata for a given mod.
+*
+* Because the mods in QuiltLoader only reference fully loaded mods, this method can be used during the mod loading process
+* to get the metadata for any candidates (tentative or not) that were present prior to this cycle.
+*/
+ModMetadata getMetadata(String modId);
```

Comment:
At this point in the loading process we are working with a case where multiple candidates for the mod may exist. Is returning one metadata correct there?

Reply:
There should only be one candidate for a given mod at this point. This method retrieves metadata for any mods that were added in a *previous* cycle, after sat4j verifies them and discards any superfluous candidates.